facilitator/src/transport/s3.rs

```diff
@@ -420,102 +421,59 @@ mod tests {
 
     const TEST_BUCKET: &str = "fake-bucket";
     const TEST_KEY: &str = "fake-key";
+    const TEST_REGION: &str = "fake-region";
 
-    fn is_create_multipart_upload_request(request: &SignedRequest) {
+    fn has_header(header_name: &str) -> Matcher {
+        let regex_text = format!("(^|&){}=", regex::escape(header_name));
+        Matcher::Regex(regex_text)
+    }
+
+    fn mock_create_multipart_upload_request() -> Mock {
         // https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html
-        assert_eq!(
-            request.method, "POST",
-            "expected CreateMultipartUpload request, found {:?}",
-            request
-        );
-        assert!(
-            request.params.contains_key("uploads"),
-            "expected CreateMultipartUpload request, found {:?}",
-            request
-        );
+        mock("POST", Matcher::Any).match_query(has_header("uploads"))
     }
 
-    fn is_upload_part_request(request: &SignedRequest) {
+    fn mock_upload_part_request() -> Mock {
         // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
-        assert_eq!(
-            request.method, "PUT",
-            "expected UploadPart request, found {:?}",
-            request
-        );
-        assert!(
-            request.params.contains_key("partNumber"),
-            "expected UploadPart request, found {:?}",
-            request
-        );
-        assert!(
-            request.params.contains_key("uploadId"),
-            "expected UploadPart request, found {:?}",
-            request
-        );
+        mock("PUT", Matcher::Any).match_query(Matcher::AllOf(vec![
+            has_header("partNumber"),
+            has_header("uploadId"),
```

Comment:
I think it's a good idea to address now. I expanded all of the `mock_*_request` functions to match expected values, and where relevant set them up to return a mocked "successful" body. The only matching I'm not really happy with is in `mock_complete_multipart_upload_request`: I match the request body by checking that the individual part ETags are there. I don't check that the entire body is well-formed XML, because I'm not sure that can be expressed with Mockito's matchers--there is no "run an arbitrary matching predicate" matcher, and I do not want to try to match XML using a regex. I think this is an OK compromise.

BTW, I noticed that the `multipart_upload` test is mocking out & then checking that calling `complete_upload` multiple times causes multiple `CompleteMultipartUpload` requests to be made (with the second-and-following requests having empty part lists). Is that behavior actually desirable? Assuming I'm not missing something, maybe it's better to consider that behavior "unspecified" (similar to iterating past the end of an iterator or polling an already-successfully-polled future) and drop that part of the test -- WDYT?

Reply:
I think your handling of the body is good. Really, we rely on `rusoto` to correctly construct XML request bodies and parse XML response bodies, so there's not that much value in us validating all that independently of the `rusoto` crate.

re: `multipart_upload` test: I think you're right, in that this is an inappropriate way to test the behavior we really care about, which is that calling `complete_upload` after the buffer has already been flushed by an `upload_part` call won't cause any further `UploadPart` API requests. I think at the time, I was being lazy and relying on implementation details of `S3Transport` to exercise that behavior. I think the right thing to do would be to add a new test case that:

- creates a multipart upload
- does a `write_all` aligned to the `minimum_upload_part_size`
- verifies that an `UploadPart` API request goes out
- does `complete_upload
- verifies that no additional `UploadPart`API request goes out

This seems like an opportune moment to write that new test, but at the same time this is code that I don't think will change much in the future so there's low risk of regressions. If we ever revisit this code, it would be to delete it and adopt the [official Rust AWS SDK](https://github.com/awslabs/aws-sdk-rust) when it becomes production quality. So your call if you want to spend the time now on that test case.